app-misc/calamares-config-livecd: added 3.3.14

StefanCristian opened the pull request:
* Branding is a prototype taken from gentoo artwork packages
* Version 3.3.14 to match the Calamares latest
* Links to sources will change, based on who is willing to handle a separate branding artwork for Gentoo.
* Kernel versioning for dracut image creation is handled manually by updating bootloader.conf and dracut.conf Will have to provide a new method of detecting these version directly from the system. But for the moment they are only prototypes and for testing purpose.
* Removing any package installed after calamares is installed will fail due to "emerge -C" requiring make.profile set.
* You cannot create user called "gentoo", because it is already created on the LiveGUI rootfsbase. This is to be discussed.

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.
- [x] I will change my identity after this PR is merged.

StefanCristian commented:
CC @ceamac @immolo .

This is it.

StefanCristian commented:
Very important note: **it does not** have kernel detection for the moment. I will have to implement that.
For testing purposes, I have used the version from the current 20.04.2025 Gentoo LiveGUI, meaning it's hardcoded within **dracut.conf** and **bootloader.conf** as version **6.12.21** and **6.12.21-dist** where applies.

StefanCristian commented:
Pre install and post install screenshots:
![image](https://github.com/user-attachments/assets/f850ef15-9b15-4d42-b600-771bccfefe69)
![image](https://github.com/user-attachments/assets/d2872805-ca21-4d1a-93ee-8a68823a7331)

StefanCristian commented:
We have a prototype Stage3 installation, with hardcoded download URL for only stage3 amd64 openrc (for now, fastest I could develop for a prototype test), and hardcoded kernel value which points to the latest stable 6.12.21-dist (also present on LIVE).

The current commit **DOES NOT** have binrepos activated, so that means it will **compile** grub and gentoo-kernel-bin before full installation :warning: 
In the next commit I plan to activate the binrepos for faster installation.

POV post-install of stage3 amd64 openrc:
![Gentoo_Stage3_CompileInstall](https://github.com/user-attachments/assets/fbefa00b-8a32-429f-a894-820c24bd1155)

thesamesam commented on app-misc/calamares-config-livecd/calamares-config-livecd-3.3.14.ebuild:
```diff
@@ -0,0 +1,47 @@
+	doins -r "${FILESDIR}/modules/"
+	doins -r "${FILESDIR}/settings.conf"
+
+	insinto /usr/lib64/calamares/modules/
```
```suggestion
	insinto /usr/$(get_libdir)/calamares/modules/
```

StefanCristian replied:
> 

this was definitely a reflex. apologies.

thesamesam commented on app-misc/calamares-config-livecd/calamares-config-livecd-3.3.14.ebuild:
```diff
@@ -0,0 +1,47 @@
+
+	insinto /etc/calamares/branding/gentoo_branding
+	doins -r "${FILESDIR}/artwork/"*
+	for i in {1..10}; do
```
```suggestion
	local i
	for i in {1..10}; do
```

thesamesam commented on app-misc/calamares-config-livecd/files/artwork/branding.desc:
```diff
@@ -0,0 +1,27 @@
+    productName:         Gentoo Linux
+    shortProductName:    Gentoo
+    version:             2.17
+    shortVersion:        2.17
```
Note that the version comes from the PV of baselayout.

thesamesam commented on app-misc/calamares-config-livecd/files/modules/downloadstage3/main.py:
```diff
@@ -0,0 +1,81 @@
+    libcalamares.job.setprogress(percent / 2)
+
+def run():
+    url = "https://distfiles.gentoo.org/releases/amd64/autobuilds/20250420T121009Z/stage3-amd64-openrc-20250420T121009Z.tar.xz"
```
This should verify the signature of the file.

StefanCristian replied:
> This should verify the signature of the file.

noted. good point.

thesamesam commented on app-misc/calamares-config-livecd/files/modules/downloadstage3/module.desc:
```diff
@@ -0,0 +1,7 @@
+type:       "job"
+name:       "downloadstage3"
+interface:  "python"
+script:     "main.py"
```
missing EOF

thesamesam commented on app-misc/calamares-config-livecd/files/modules/dracut.conf:
```diff
@@ -0,0 +1,14 @@
+
+# Optional: define a list of strings to be passed as arguments to Dracut
+# By default, -f is always included
+options: [ "-H", "-f", "-o systemd", "-o systemd-initrd", "-o systemd-networkd", "-o dracut-systemd", "-o plymouth", "--early-microcode", "--kver=6.12.21-gentoo-dist" ]
```
missing EOF

thesamesam commented on app-misc/calamares-config-livecd/files/modules/dracut.conf:
```diff
@@ -0,0 +1,14 @@
+# If you want to specify another filename for the resulting image,
+# set a custom name, including the path
+#
+initramfsName: /boot/initramfs-6.12.21-gentoo-dist.img
```
Please add TODOs for stuff like this

thesamesam reviewed (commented):
Left some small comments for things I noticed.

StefanCristian commented:
- Corrected the kernel detection mechanism
- Added PV-detection from baselayout in order to determine the Gentoo Linux version
- Corrected the autodownloader to be fixed on latest version found in https://distfiles.gentoo.org/releases/amd64/autobuilds/current-stage3-amd64-openrc/, whichever that will be
- Added checksum verification for downloaded tar.xz
- :warning: issue remains: we need a precompiled gentoo-kernel-bin, since the chaindeps take a lot of time to build on VM

For prototype testing purposes, it installs a Stage3 AMD64 OpenRC succesfully, bootable.
TODO:
- Menu for choosing Stage3s

ceamac commented on app-misc/calamares-config-livecd/calamares-config-livecd-3.3.14.ebuild:
```diff
@@ -0,0 +1,63 @@
+    default
+
+    local baselayout_version
+    baselayout_version="$(best_version sys-apps/baselayout | sed -E 's:.*/[^-]+-::')"
```
This looks quite fragile.  How about something like this:
```
baselayout_version=$(best_version sys-apps/baselayout)
baselayout_version=${baselayout_version#*layout-}
```

StefanCristian replied:
> This looks quite fragile. How about something like this:
> 
> ```
> baselayout_version=$(best_version sys-apps/baselayout)
> baselayout_version=${baselayout_version#*layout-}
> ```

definitely much better

ceamac commented on app-misc/calamares-config-livecd/calamares-config-livecd-3.3.14.ebuild:
```diff
@@ -0,0 +1,63 @@
+
+    cp "${FILESDIR}/artwork/branding.desc" "${WORKDIR}/branding.desc" || die
+
+    if [[ -n "${baselayout_version}" ]]; then
```
No need for quotes here

ceamac commented on app-misc/calamares-config-livecd/files/modules/bootloader.conf:
```diff
@@ -0,0 +1,27 @@
+# efiBootloaderId: "dirname"
+grubProbe: "grub-probe"
+efiBootMgr: "efibootmgr"
+installEFIFallback: true
```
No eof

ceamac commented on app-misc/calamares-config-livecd/calamares-config-livecd-3.3.14.ebuild:
```diff
@@ -0,0 +1,63 @@
+
+	local i
+	for i in {1..10}; do
+		newins gentoo-livecd-2007.0/800x600.png "${i}.png"
```
Out of curiosity, what is this supposed to do?

StefanCristian replied:
> Out of curiosity, what is this supposed to do?

just replacing the background images from 1 to 10 with the (testing) old background images, most probably someone will have to replace that artwork package to look nice.

I propose this new one:
![Gentoo Linux 2025 4 Design](https://github.com/user-attachments/assets/f04272c0-d21f-457d-ad78-12bcf298c493)

ceamac reviewed (commented):
Some quick notes, I only looked a bit, didn't test it or something

StefanCristian commented:
* Added support for choosing stage3 installation
* Only issue with installation is musl-hardened, due to lack of /usr/share/zoneinfo package (can be installed in a earlier step)

The idea behind is choosing which stage3 fits the user best, and the downloader will download the latest current version of the chosen option, unpacks it, installs the kernel and grub, and generates the initramfs via dracut (for the moment).
It has no special post-installation configuration for now.

Screenshot with the menu:
![image](https://github.com/user-attachments/assets/d51b701a-5a18-49ea-b422-173d09fd46e6)

ceamac commented on app-misc/calamares-gentoo-livecd/calamares-gentoo-livecd-3.3.14_p20250915-r1.ebuild:
```diff
@@ -0,0 +1,40 @@
+
+inherit desktop python-single-r1
+
+DESCRIPTION="Gentoo Linux ${MY_PN} installer configuration for LiveCD"
```
I wouldn't use ${MY_PN} here.  It's not forbidden per se, but it's ugly and not bringing any value

ceamac commented on app-misc/calamares-gentoo-livecd/calamares-gentoo-livecd-3.3.14_p20250915-r1.ebuild:
```diff
@@ -0,0 +1,40 @@
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~amd64"
```
Blank line here please

ceamac commented on app-misc/calamares-gentoo-livecd/calamares-gentoo-livecd-3.3.14_p20250915-r1.ebuild:
```diff
@@ -0,0 +1,40 @@
+RDEPEND="${DEPEND}
+		>=app-admin/${MY_PN}-3.3.14-r2[livecd]"
+
+REQUIRED_USE="${PYTHON_REQUIRED_USE}"
```
Please move this below KEYWORDS (REQUIRED_USE should be after IUSE, which is after KEYWORDS)

ceamac commented on app-misc/calamares-gentoo-livecd/calamares-gentoo-livecd-3.3.14_p20250915-r1.ebuild:
```diff
@@ -0,0 +1,40 @@
+
+REQUIRED_USE="${PYTHON_REQUIRED_USE}"
+
+S="${WORKDIR}/${P}"
```
Please remove S if it is the default (`${WORKDIR}/${P}"`)

ceamac commented on app-misc/calamares-gentoo-livecd/calamares-gentoo-livecd-3.3.14_p20250915-r1.ebuild:
```diff
@@ -0,0 +1,40 @@
+
+DESCRIPTION="Gentoo Linux ${MY_PN} installer configuration for LiveCD"
+HOMEPAGE="https://github.com/StefanCristian/calamares-gentoo-livecd"
+SRC_URI="https://github.com/StefanCristian/${PN}/archive/refs/tags/${PV}.tar.gz -> ${PF}.tar.gz
```
We usually avoid `${PF}` here, use `${P}` instead.  If upstream tarball changes, do a version bump, not a revision bump; revision bumps should reuse the same tarball

ceamac commented on app-misc/calamares-gentoo-livecd/calamares-gentoo-livecd-3.3.14_p20250915-r1.ebuild:
```diff
@@ -0,0 +1,40 @@
+		mirror://gentoo/gentoo-artwork-livecd-2007.0.tar.bz2
+		mirror://gentoo/gentoo-artwork-0.2.tar.bz2"
+
+LICENSE="GPL-2"
```
Also there is no license file in the repo, please ask upstream to add one :)  You should prefer `GPL-2+` or `GPL-3`, since calamares itself is `GPL-3`.

ceamac commented on app-misc/calamares-gentoo-livecd/calamares-gentoo-livecd-3.3.14_p20250915-r1.ebuild:
```diff
@@ -0,0 +1,40 @@
+KEYWORDS="~amd64"
+DEPEND="${PYTHON_DEPS}"
+RDEPEND="${DEPEND}
+		>=app-admin/${MY_PN}-3.3.14-r2[livecd]"
```
You don't need `>=` here since older versions don't have the `livecd` USE flag

ceamac reviewed (changes requested):
Please rebase against master to get rid of the CI error
